### PR TITLE
Milestone 3: Game logic

### PR DESCRIPTION
**Pull request by nyatindopatrick**
Created classes to encapsulate the different parts of the game logic.

### CODE REVIEW

**Review by mosesogwo — CHANGES_REQUESTED** [review]
**Hi @nyatindopatrick, @mariosknl. Good work:+1:. Some changes are required :fast_forward:**
- No **puts, print or get.chomp** outside of main.rb. Refactor your methods in such a way that your puts, prints or get.chomp requests are made from main.rb.
- When a user choice is requested and enter is pressed without making a choice, 9 is automatically selected. The game should request for the user's valid input before continuing.
- When a winning move is made, the game waits for the next move before determining that there is a win. This allows for a situation like we have below where you have two winners on the board:
desktop
Once a winning move is made, the game should report it and end.
- Place each class in a separate file within the lib folder.
- Your **win_moves** method logically fits into the board class. Move it there and make the place variable a non-global variable. You may not need a winner class.
- Your **win_moves** method should be refactored to take a player or player_symbol as argument and be called to check after each player's move instead of checking both players in one go.
- Try to avoid the use of global variables. Instead of referencing the $place variable in your update_board method, you can have a method in your board class that checks if a position in **place** is available. This method can then be used in main.rb to check the player's choice before the update_board is called.
- In the spirit of OOP, let us have a class to represent our player object. Instances of the player object can then be created at the start of the game. 

_**Kindly make these changes and submit another code review request.**_

**Any question? You can reach me on Slack (Moses Ogwo)**

**Cheers:stars:**

### Status: Changes Requested.

**Review by pelzolga123 — CHANGES_REQUESTED** [review]
Project requires changes.

**Comment by pelzolga123** [comment]
## Hi Team!  
### I see you worked on changes required previously, sadly I couldn't start game because of error:
image
 * [ ] Error pointed above should be fixed.
 * [ ] The project should use at least two classes (as previous review pointed, you can add class for player). The code of the project should be logically divided between classes. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/classes.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

### Please kindly fix the above issues and send another code review request.

If you have any questions or need help, you can message me on slack (@ Olga Pelts). *Happy coding*

File: lib/game_logic.rb [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  end
+
+  def empty_space(param2)
+    input_message = Helper.prompt_users(param2)
+    if @board[input_message - 1] == ' '
+      input_message
+    else
+      empty_space(param2)
+    end
+  end
```
Comment (shshamim63) [inline lib/game_logic.rb]: Base on the current formation, it will be tough to test it later.

File: lib/game_play.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+require_relative './helper_module.rb'
+
+class Play
+  def initialize(current_player)
+    @current_player = current_player
+  end
+
+  def player_sign(param1)
+    players_sign = @current_player == param1 ? 'X' : 'O'
+    players_sign
```
Comment (shshamim63) [inline lib/game_play.rb]: Line 10 is redundant here. Line 9 can do the trick.

File: lib/helper_module.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    end
+  end
+
+  def self.end_game(par)
+    puts par
+    puts 'Play again? (Y/N)'
+    user_in = gets.chomp.downcase
+    end_game(par) unless %w[y and n].include?(user_in)
+    user_in
+  end
```
Comment (shshamim63) [inline lib/helper_module.rb]: Definitely not in main.rb. So violates no puts/ chomp outside of main.rb instruction.

File: lib/game_logic.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    \t -------------
+    GAME
+  end
+
+  # Checking and eturning the winner
+  def win_moves(player)
+    temp = []
+    @board.each_with_index do |val, i|
+      player == @player1 ? (temp << i if val == 'X') : (temp << i if val == 'O')
+    end
```
Comment (shshamim63) [inline lib/game_logic.rb:26]: #### Optional
It provides current player moves. Maybe the moves generate a winning combination, maybe not. So name of the method does not fit with its operation. Maybe current_player_moves is more suitable. though too long.

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :space_invader: Changes Required :space_invader: 
---
Hi @nyatindopatrick and @mariosknl :wave: ,
Awesome job :clap: , The game is functional now. But there are some changes that were asked by the previous reviewer still remains uncovered. So follow my comments below to complete this project. 
___
### :spades: Mandatory Changes :spades: 
___
1. No `puts/print` and `get.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)). I can see some puts and chomp in a module. Obviously that is outside of main.rb. Now why we want these messages in main.rb? Ans: So that we can test most of our methods properly(That you will do on your next project.) On milestone two you have added all the messages needed to guide the users. All you need is some sequential and supportive methods enclosed in some classes. Kindly fix the issue.
___
##### Kindly fix the issues mentioned above and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain

**Review by Torres-ssf — APPROVED** [review]
### Project Approved :+1: :tada:

Hello there :hand: 
Great job implementing the requested changes.
**I'm approving this PR, good luck in the next milestone**

**Sergio Torres**

**Pull request merged**